### ISSUE TITLE
Support nested aggregation functions in a DDL

### ISSUE THREAD

**Issue by drealeed** [hpcc-systems/Visualization#2291]
HIPIE can now do nested aggregations like MAX(FIELD)-MIN(FIELD), SUM(field)/COUNT, etc. in roxie based aggregations. This will need to be emulated on the WUID-based dashboard side of things.

here's a proposed DDL for a wuid-based dashboard. What do you think? It would reference workunit http://10.173.147.1:8010/?Wuid=W20170828-103707&Widget=WUDetailsWidget#/stub/Summary.

The change in the DDL is in the field definition properties:
`````
                                        "function": "dynamic",
 					"label": "sum/ave",
 					"params": {
 						"param1": "mark1_SUM",
 						"param2": "/",
 						"param3": "mark1_AVE"
 					},
`````
And look like this roxie-based dashboard:
http://viz.hpccsystems.com/v1.16.0/apps/legacyDash.html?Hostname=10.173.147.1&Port=8010&Protocol=http&Wuid=W20170828-103111&ResultName=testaggs_Comp_Ins002_DDL

`````
 {
 	"dashboards": [{
 		"id": "Ins002_TestAggs",
 		"label": "TestAggs",
 		"title": "Sort Passthrough Output",
 		"visualizations": [{
 			"events": {
 				"click": {
 					"updates": [{
 						"datasource": "Ins002_dsOutput",
 						"instance": "Ins002",
 						"mappings": {
 							"subject": "subject"
 						},
 						"merge": false,
 						"visualization": "Dynamic"
 					}]
 				}
 			},
 			"fields": [{
 				"id": "subject",
 				"properties": {
 					"datatype": "string",
 					"label": "subject",
 					"type": "string"
 				}
 			}, {
 				"id": "Base_COUNT",
 				"properties": {
 					"datatype": "integer",
 					"function": "SUM",
 					"label": "Base_COUNT",
 					"params": {
 						"param1": "Base_COUNT"
 					},
 					"type": "integer"
 				}
 			}],
 			"id": "Subjects",
 			"label": ["subject", "Base_COUNT"],
 			"source": {
 				"id": "Ins002_dsOutput",
 				"mappings": {
 					"value": ["subject", "Base_COUNT"]
 				},
 				"output": "View_Subjects"
 			},
 			"title": "subjects",
 			"type": "TABLE"
 		}, {
 			"fields": [{
 				"id": "subject",
 				"properties": {
 					"datatype": "string",
 					"label": "subject",
 					"type": "string"
 				}
 			}, {
 				"id": "mark1_SUM",
 				"properties": {
 					"datatype": "real",
 					"function": "SUM",
 					"label": "sum mark 1",
 					"params": {
 						"param1": "mark1_SUM"
 					},
 					"type": "real"
 				}
 			}, {
 				"id": "mark1_MIN",
 				"properties": {
 					"datatype": "integer8",
 					"function": "MIN",
 					"label": "min mark 1",
 					"params": {
 						"param1": "mark1_MIN"
 					},
 					"type": "integer8"
 				}
 			}, {
 				"id": "mark1_MAX",
 				"properties": {
 					"datatype": "integer8",
 					"function": "MAX",
 					"label": "max mark 1",
 					"params": {
 						"param1": "mark1_MAX"
 					},
 					"type": "integer8"
 				}
 			}, {
 				"id": "mark1_AVE",
 				"properties": {
 					"datatype": "real",
 					"function": "AVE",
 					"label": "average mark 1",
 					"params": {
 						"param1": "mark1_SUM",
 						"param2": "Base_COUNT"
 					},
 					"type": "real"
 				}
 			}, {
 				"id": "L_mark1_MAX_MINUS_mark1_MIN_R_COMPUTED",
 				"properties": {
 					"datatype": "real",
 					"function": "DYNAMIC",
 					"label": "range mark1",
 					"params": {
 						"param1": "mark1_MAX",
 						"param2": "-",
 						"param3": "mark1_MIN"
 					},
 					"type": "real"
 				}
 			}, {
 				"id": "L_mark1_SUM_DIVIDE_Base_COUNT_R_COMPUTED",
 				"properties": {
 					"datatype": "real",
 					"function": "DYNAMIC",
 					"label": "sum/count",
 					"params": {
 						"param1": "mark1_SUM",
 						"param2": "/",
 						"param3": "Base_COUNT"
 					},
 					"type": "real"
 				}
 			}, {
 				"id": "L_mark1_SUM_DIVIDE_mark1_AVE_R_COMPUTED",
 				"properties": {
 					"datatype": "real",
 					"function": "dynamic",
 					"label": "sum/ave",
 					"params": {
 						"param1": "mark1_SUM",
 						"param2": "/",
 						"param3": "mark1_AVE"
 					},
 					"type": "real"
 				}
 			}],
 			"id": "Dynamic",
 			"label": ["subject", "sum mark 1", "min mark 1", "max mark 1", "average mark 1", "range mark1", "sum/count", "sum/ave"],
 			"source": {
 				"id": "Ins002_dsOutput",
 				"mappings": {
 					"value": ["subject", "mark1_SUM", "mark1_MIN", "mark1_MAX", "mark1_AVE", "L_mark1_MAX_MINUS_mark1_MIN_R_COMPUTED", "L_mark1_SUM_DIVIDE_Base_COUNT_R_COMPUTED", "L_mark1_SUM_DIVIDE_mark1_AVE_R_COMPUTED"]
 				},
 				"output": "View_Dynamic"
 			},
 			"type": "TABLE"
 		}]
 	}],
 	"datasources": [{
 		"WUID": true,
 		"filter": [{
 			"fieldid": "subject",
 			"nullable": false,
 			"rule": ">"
 		}],
 		"id": "Ins002_dsOutput",
 		"outputs": [{
 			"from": "Ins002_dsOutput_View_Subjects",
 			"id": "View_Subjects",
 			"notify": ["Subjects"]
 		}, {
 			"filter": [{
 				"fieldid": "subject",
 				"nullable": false,
 				"rule": ">"
 			}],
 			"from": "Ins002_dsOutput_View_Subjects",
 			"id": "View_Dynamic",
 			"notify": ["Dynamic"]
 		}]
 	}],
 	"hipieversion": "1.9.0",
 	"visualizationversion": "<ECLSTRINGBREAK> + $.HIPIEConfig.VizVersion + <ECLSTRINGBREAK>"
 }
`````

**Comment by GordonSmith** [comment]
@drealeed What is the timeline for this?  Can I push it off into v2.0 dashboards?

**Comment by drealeed** [comment]
Yes. I just tried to create a situation in HIPIE where this would be a problem and could not. Currently, because filter fields are required to be included in the widgets being filtered upon, and because of HIPIE's rules in how many named outputs to create for certain widgets, in all cases I tested the named outputs are being created at the level of aggregation needed by a widget, with no need to further roll up the data using the functions.
